### PR TITLE
Ravil/spack

### PR DESCRIPTION
**Pull request by ravil-mobile**
Automatic installation of SeisSol dependencies with Spack

### CODE REVIEW

File: Documentation/installation-with-spack.rst [hunk 1]
```diff
@@ -0,0 +1,391 @@
+Installation with Spack
+=======================
+
+Installation of any HPC application can be tricky. SeisSol is not an exception. 
+To considerably alleviate the installation process, we provide few scripts which 
+relies on `Spack <https://github.com/spack/spack/wiki>`_, which is a new HPC
```
Comment (sebwolf-de) [inline Documentation/installation-with-spack.rst]: rely

Comment (ravil-mobile) [reply Documentation/installation-with-spack.rst]: thanks!

File: Documentation/installation-with-spack.rst [hunk 2]
```diff
@@ -0,0 +1,391 @@
...
+Examples
+~~~~~~~~
+
+.. code-block:: bash
+
+  # 1. with intel compiler suite
+  spack install seissol-env +mpi +asagi ^intel-mpi@2018.2.199 %intel@17.0.2
+
+  # 2. with gcc compiler suite
+  spack install seissol-env +mpi +asagi ^openmpi@3.1.5 %gcc@8.3.0
```
Comment (sebwolf-de) [inline Documentation/installation-with-spack.rst]: When I execute this line (tried on LRZ workstation and personal notebook) I get the following error:
seissol-env requires cmake version 3.12.0:3.16.2, but spec asked for 3.16.5
I have installed cmake 3.15.2 and 3.16.4 respectively.

File: Documentation/installation-with-spack.rst [hunk 3]
```diff
@@ -0,0 +1,391 @@
...
+.. code-block:: bash
+
+  module load seissol-env-develop-<compiler>-<hash>
+
+  # if you compile seissol-env with a compiler installed with Spack
+  # you may be required to load that compiler as well
+  module load <compiler>
+
+
+After that, you can compiler SiesSol using either CMake or
```
Comment (sebwolf-de) [inline Documentation/installation-with-spack.rst]: typo

File: Documentation/installation-with-spack.rst [hunk 4]
```diff
@@ -0,0 +1,391 @@
...
+
+
+After that, you can compiler SiesSol using either CMake or 
+:ref:`Scons <compiling-seissol>`.
+
+
+SeisSol-Utils
+-------------
+
+By default, the scrip installs:
```
Comment (sebwolf-de) [inline Documentation/installation-with-spack.rst]: typo

File: Documentation/installation-with-spack.rst [hunk 5]
```diff
@@ -0,0 +1,391 @@
...
+
+- pumgen (without a Simmetrix support)
+- gmsh (without a GPU support)
+- gmsh2gambit
+- cube_c
+- rconv
+- SiesSol Cookbook, which contains some examples to run
```
Comment (sebwolf-de) [inline Documentation/installation-with-spack.rst]: typo

File: Documentation/installation-with-spack.rst [hunk 6]
```diff
@@ -0,0 +1,391 @@
...
+  cd $BENCHMAKRS
+
+
+
+Tips and Tricks
+---------------
+
+.. _spack_known_issues:
+
+1. Spack builds the entire dependency graph before compaling and installing.
```
Comment (sebwolf-de) [inline Documentation/installation-with-spack.rst]: typo

File: Documentation/installation-with-spack.rst [hunk 7]
```diff
@@ -0,0 +1,391 @@
...
+1. Spack builds the entire dependency graph before compaling and installing. 
+The graph includes all libs and packages which are necessary to build your 
+application, including packages like: *tar, gzip, zlib,  autoconf, 
+cmake, automake, pkgconf, m4, ncurses, etc*. Packages like these do not 
+affect performance of your application but help Spack to install it. 
+Therefore, it is not necessary to install them again and again. You can 
+install such  packages only once and mark them as Default 
+`(External) <https://spack-tutorial.readthedocs.io/en/latest/tutorial_configuration.html#external-packages>`_.
+and Non-Buildable. It can speed-up installation of SeisSol-Env and SeisSol-utils 
+considerably. You will need to modify and edit **~/.spack/packages.yaml** file.
```
Comment (sebwolf-de) [inline Documentation/installation-with-spack.rst]: Wouldnt it be a good idea to reverse the process here? Make it the default to not build all these basic packages and tell people to also build tar, gzip, ... when they encounter problems.

File: Documentation/installation-with-spack.rst [hunk 8]
```diff
@@ -0,0 +1,391 @@
...
+
+
+Known Issues
+------------
+
+1. Spack is a really live project with dozens of commits per day. It is 
+diffucult for us to keep the same pace with Spack. A new version of Spack
```
Comment (sebwolf-de) [inline Documentation/installation-with-spack.rst]: typo

File: Documentation/installation-with-spack.rst [hunk 9]
```diff
@@ -0,0 +1,391 @@
...
+may not work because of new added features what we may not be aware of. 
+Therefore, it may be necessary to use an older version of Spack. You
+can simply do it by moving the HEAD of your locally installed Spack
+repository to an old commit:
+
+.. code-block:: bash
+
+  cd $SPACK_ROOT
+  git checkout <a previous SPACK commit>
```
Comment (sebwolf-de) [inline Documentation/installation-with-spack.rst]: Maybe provide todays spack commit hash?

File: spack_support/packages/seissol-env/package.py [hunk 10]
```diff
@@ -0,0 +1,117 @@
+# Copyright 2013-2020 Lawrence Livermore National Security, LLC and other
```
Comment (sebwolf-de) [inline spack_support/packages/seissol-env/package.py]: I guess this file is written by you, Ravil, based on some sample taken from Lawrence Livermore? So maybe alter the LICENCE header accordingly. But I am no expert on legal stuff.

File: spack_support/packages/seissol-utils/package.py [hunk 11]
```diff
@@ -0,0 +1,108 @@
+# Copyright 2013-2020 Lawrence Livermore National Security, LLC and other
```
Comment (sebwolf-de) [inline spack_support/packages/seissol-utils/package.py]: Same as for seissol-env

File: spack_support/packages/seissol-env/package.py (hunk 10 above)
Comment (ravil-mobile) [reply spack_support/packages/seissol-env/package.py]: It is the standard license agreement enforced by spack. Basically, this line is automatically generated by spack when one creates an installation script. I checked other scripts from hdf5 and openmpi. It is the same there

File: Documentation/installation-with-spack.rst (hunk 7 above)
Comment (ravil-mobile) [reply Documentation/installation-with-spack.rst]: As far as I know there is only one way how to force spack to not build some packages: to add them to your ~/.spack/packages.yaml and mark them as Non-Buildable

Comment (ravil-mobile) [reply Documentation/installation-with-spack.rst]: I would like to do the same as you suggested but I couldn't find a solution to enforce spack to prevent building these packages inside of an installation script

File: Documentation/installation-with-spack.rst (hunk 9 above)
Comment (ravil-mobile) [reply Documentation/installation-with-spack.rst]: Honestly, I am not sure about it. I cannot check all possible combinations with respect to all possible installation options. Therefore, I am not sure that today's snippet of Spack is a perfect reference

File: Documentation/installation-with-spack.rst [hunk 12]
```diff
@@ -0,0 +1,392 @@
...
+*NOTE*: mpi is a virtual package, a user must specify a concrete implementation
+of the standard
+
+Examples
+~~~~~~~~
+
+.. code-block:: bash
+
+  # 1. with intel compiler suite
+  spack install seissol-env +mpi +asagi %intel@17.0.2 ^intel-mpi@2018.2.199
```
Comment (sebwolf-de) [inline Documentation/installation-with-spack.rst]: Shouldn't we include the ^cmake@3.16.0 part after our discussion on slack?

File: Documentation/installation-with-spack.rst (hunk 7 above)
Comment (sebwolf-de) [reply Documentation/installation-with-spack.rst]: Okay, so let's leave it like this.

File: Documentation/installation-with-spack.rst (hunk 9 above)
Comment (sebwolf-de) [reply Documentation/installation-with-spack.rst]: Okay, I see there are just to many options to try all of them. But maybe we can add something like: On SuperMUC commit xxxxxx works fine for the combinations described above.

Comment (ravil-mobile) [reply Documentation/installation-with-spack.rst:350]: It is a good idea. Additionally, we can provide package.yaml file with not-buildable packages (an use it as an exaple) and a specify a particular commit for Spack. Let me see whether I can do it for SuperMUC because the admins did whatever they could to prevent compiling on production nodes but maybe i will manage to find a way around

**Comment by ravil-mobile** [comment]
Hi all,

One of the last issues before the merge was a request to test the installation of NG. Finally I've got access to NG. I talked to one of the admins about how to use Spack there from the user's perspective. The short answer was to wait and play with Spack on the Linux cluster.

The NG is going to be on maintenance and it will support Spack@15 in the future (but nobody knows when). The LRZ people use LinuxCluster as a playground for Spack. In other words, they test how users interact with Spack, collect issues and make notes about what should be done on the NG in order to fully support Spack for users like us. They told me that everything would work more or less the same on NG as it is on the Linux Cluster right now. 

So, I will try to install everything on the Linux Cluster and document. The best strategy with respect to NG is just to wait.

One question that I wanted to ask is: maybe we can move Spack support to a separate repository? What do you think?

**Comment by Thomas-Ulrich** [comment]
Can you just install and use your own spack on supermuc? E.g. this is 
what I used for installing python vtk on NG:git clone 
git@github.com:spack/spack.git
cd spack/bin
#load all the required compilers, e.g. module load gcc
./spack compilers
#Install spack on the computer with internet access, then download the 
required packages using:
./spack mirror create -d spack-mirror-2019-10-17/ -D 
vtk+python^hdf5+hl+mpi^python@3.6.0 py-numpy^python@3.6.0 
py-lxml^python@3.6.0 py-h5py^python@3.6.0
#copy spack-mirror-2019-10-17/ to supermuc and then run:
./spack mirror add mvtk spack-mirror-2019-10-17/
./spack install py-numpy^python@3.6.0 py-lxml^python@3.6.0 
py-h5py^python@3.6.0 vtk+python^hdf5+hl+mpi^python@3.6.0
./spack activate py-numpy^python@3.6.0 py-lxml^python@3.6.0 
py-h5py^python@3.6.0
#Here are other python modules that might be interesting to install: 
py-matplotlib^python@3.6.0 py-scipy^python@3.6.0

**Comment by ravil-mobile** [comment]
Hi @Thomas-Ulrich,

Yes, you can.  I guess you ended up with Spack installing lots of low-level stuff (even m4, tar, zlib). 

However, it should not be like this in a cluster. Admins of a cluster "mark" all these low-level packages as well as compilers and different mpi installations only once "in a system level" which becomes a default setup for all users. Additionally, the admins setup default paths to some caches: for builds and sources, as well as a tmp folder for Spack.

For example, one can load Spack on the linux cluster as following
```
module load user_spack
```
which will set up SPACK_ROOT to /lrz/sys/spack/user/release/20.1/spack
there is a folder called config with (as the name suggest) all Spack configurations for the linux cluster.
The most interesting is *upstreams.yaml* and a filed inside called *install_tree*

Basically, it say where spack should look for a package first. If it can find a binary there it will just copy it to my local spack folder. Otherwise, spack will install the package from sources. It will look at user's defined mirrors (basically what you proposed) and try to find sources. If Spack cannot find source files there. It will try to fetch it from the Internet.

**Comment by ravil-mobile** [comment]
Your solution is fine because there is no official Spack support on NG. I can try to hack the Spack that they use but I don't think that is a good idea. Also I can try to setup Spack for a project but this solution is not portable between projects. Maybe it is better to wait till they officially provide Spack support. What do you think?

**Comment by brietzke** [comment]
optimal would be to bring seissol/package.py to upstream spack

**Comment by ravil-mobile** [comment]
Hi, thanks for joining our conversation. Yes, you right. But I don't know whether spack reviewers will agree upon this style: https://github.com/SeisSol/SeisSol/blob/e4602bede0aab8509f59b019db94d8889152b267/spack_support/packages/seissol-env/package.py#L68-L105

Basically Im looping through all dependencies and add them to the environment variables. It is convenient for me as an user because I eventually need to load only one module. However, I guess it is not what spack developers want a script to perform. But on another hand, it is a `BundlePackage`.

Before up-streaming seissol-env and seissol-utils I need to upstream other packages (you can have a look here: https://github.com/SeisSol/SeisSol/tree/ravil/spack/spack_support/packages). Some of them are not ours. Do you have any idea what I should do in this case? For example, what should I do with `cxxtest` and `impalajit`?

Thanks in advance

File: Documentation/installation-with-spack.rst [hunk 13]
```diff
@@ -0,0 +1,392 @@
...
+Additionally, a user can customize each individual dependency using 
+Spack 
+`recursive syntax <https://spack.readthedocs.io/en/latest/basic_usage.html#specs-dependencies>`_. 
+
+
+Options
+~~~~~~~
+
+- *asagi* [**default=on**, off] - installs asagi 
+- *building_tools* [**default=on**, off] - installs scons and cmake
```
Comment (sebwolf-de) [inline Documentation/installation-with-spack.rst:218]: Here you say, default=on, above you reason that scons and cmake are optional.

File: Documentation/installation-with-spack.rst [hunk 14]
```diff
@@ -0,0 +1,392 @@
...
+
+  module load seissol-env-develop-<compiler>-<hash>
+
+  # if you compile seissol-env with a compiler installed with Spack
+  # you may need to load that compiler as well
+  module load <compiler>
+
+
+After that, you can compile SeisSol using either CMake or 
+:ref:`Scons <compiling-seissol>`.
```
Comment (sebwolf-de) [inline Documentation/installation-with-spack.rst:260]: compiling-seissol has now cmake as the default.

File: Documentation/installation-with-spack.rst [hunk 15]
```diff
@@ -0,0 +1,392 @@
...
+
+
+SeisSol-Utils
+-------------
+
+By default, the script installs:
+
+- pumgen (without a Simmetrix support)
+- gmsh (without a GPU support)
+- gmsh2gambit
```
Comment (sebwolf-de) [inline Documentation/installation-with-spack.rst:270]: With the new gmsh funcionality of pumgen, gmsh2gambit might be obsolete.

File: Documentation/installation-with-spack.rst [hunk 16]
```diff
@@ -0,0 +1,392 @@
...
+- cube_c
+- rconv
+- SeisSol Cookbook, which contains some examples to run
```
Comment (sebwolf-de) [inline Documentation/installation-with-spack.rst:273]: SeisSol Cookbook is now SeisSol/Examples, I guess.

File: Documentation/installation-with-spack.rst [hunk 17]
```diff
@@ -0,0 +1,392 @@
...
+
+As in case of *seissol-env*, you need **scons** and, therefore, **python3** for 
+compiling. However, installation of these packages is optional to save 
+the set-up time.
+
+
+Options
+~~~~~~~
+
+- *benchmarks* [on, **default=off**] - installs SeisSol benchmarks. Make sure that you have access to the SeisSol LRZ-gitlab account.
```
Comment (sebwolf-de) [inline Documentation/installation-with-spack.rst:283]: Also here: Everything is merged in SeisSol/Examples.

File: Documentation/installation-with-spack.rst [hunk 18]
```diff
@@ -0,0 +1,392 @@
...
+
+.. code-block:: bash
+
+  module load seissol-utils-develop-<compiler>-<hash>
+
+  # to access the Cookbook
+  cd $COOKBOOK
+
+  # to access the Benchmakrs
+  cd $BENCHMAKRS
```
Comment (sebwolf-de) [inline Documentation/installation-with-spack.rst:316]: Also this is just SeisSol/Examples now

File: spack_support/packages/seissol-utils/package.py [hunk 19]
```diff
@@ -0,0 +1,154 @@
...
+
+    variant('cookbook', 
+            default=False, 
+            description="fetches cookbook. Be sure that you "
+                        "have access to the SeisSol LRZ-gitlab repo.")
+
+    variant('benchmarks', 
+            default=False, 
+            description="fetches benchmarks. Be sure that you "
+                        "have access to the SeisSol LRZ-gitlab repo.")
```
Comment (sebwolf-de) [inline spack_support/packages/seissol-utils/package.py:34]: Also here: Move to SeisSol/Examples.

File: spack_support/packages/seissol-utils/package.py [hunk 20]
```diff
@@ -0,0 +1,154 @@
...
+
+    resource(name='benchmarks', 
+             git='https://gitlab.lrz.de/seissol.org/benchmarks',
+             when='+benchmarks', 
+             placement='benchmarks')
+
+    depends_on("hdf5@1.8.21 +fortran +shared +mpi")
+    depends_on('netcdf-c@4.4.0 +shared +mpi')
+    depends_on('pumgen')
+    depends_on('glm@0.9.7.1')
```
Comment (sebwolf-de) [inline spack_support/packages/seissol-utils/package.py:52]: We removed glm recently.

File: spack_support/packages/seissol-utils/package.py [hunk 21]
```diff
@@ -0,0 +1,154 @@
...
+                cmake(".")
+                make()
+
+    def install(self, spec, prefix):
+
+        if "+cookbook" in spec:
+            install_tree("cookbook", prefix.cookbook)
+
+        if "+benchmarks" in spec:
+            install_tree("benchmarks", prefix.benchmarks)
```
Comment (sebwolf-de) [inline spack_support/packages/seissol-utils/package.py:114]: You already know, what I mean here ;-)

**Review by sebwolf-de — COMMENTED** [review]
Hi, good steps forward. I guess a lot of people already use spack-support successfully. So IMO we can move this to another repository as Ravil supposed and think about upstreaming this there.
